fig.go

```diff
@@ -246,12 +313,23 @@ func (f *fig) formatEnvKey(key string) string {
 	return strings.ToUpper(key)
 }
 
-// setDefaultValue calls setValue but disallows booleans from
-// being set.
+// setDefaultValue calls setValue unless the value satisfies the
+// StringUnmarshaler interface or is of a boolean type.
 func (f *fig) setDefaultValue(fv reflect.Value, val string) error {
 	if fv.Kind() == reflect.Bool {
 		return fmt.Errorf("unsupported type: %v", fv.Kind())
 	}
+	if reflect.PointerTo(fv.Type()).Implements(reflect.TypeOf((*StringUnmarshaler)(nil)).Elem()) {
+		vi := reflect.New(fv.Type()).Interface()
+		if unmarshaler, ok := vi.(StringUnmarshaler); ok {
+			err := unmarshaler.UnmarshalString(val)
+			if err != nil {
+				return err
+			}
+			fv.Set(reflect.ValueOf(vi).Elem())
+		}
+		return nil
```

Comment:
Instead of returning `nil` we should continue with the regular flow so that it's consistent with the behaviour in the decoder hook.

Reply:
If we continue the workflow it will run through the `switch fv.Kind()` statement and it will match the `case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64`, which then fail trying to parse the string as uint `strconv.ParseUint`

Reply:
It might be the case that an `int` is provided (e.g. 1 for `ListenerTCP`), in which case it would not fail. However, given that the type satisfies the `StringMarshaler` interface, imo it's fine to have the string representation take precedence over the actual representation. So I agree that continuing the workflow might not be the best approach.

That being said, line 341 should not be reachable given the check on line 332, so returning an error would be more appropriate as opposed to just doing nothing. 

WDYT? (If you agree, let's make the same modification in the decode hook as well.)

Reply:
Line 341 will always be reached since we only do a `fv.Set` in line 329 but do nothing else. If we error out here, it will always return that error if the `StringMarshaler` is satisfied.

Reply:
We could move the `return nil` after the `fv.Set()` to only return if the `StringMarshaler` was executed successfully and then return with an error in in 341, to make sure that if that line is ever reached, it throws an error.

Reply:
You're right, I thought 339 was returning. Your suggestion LGTM.

Reply:
Awesome, thanks for all the great feedback.